lib/libgcov/gcov_gcc.c

```diff
@@ -0,0 +1,411 @@
+#define NB_COUNTERS 10
+#elif GCC_VERSION_7 <= __GCC_VERSION && __GCC_VERSION < GCC_VERSION_10
+/* 56b653f1a37372ceaba9ec6cbfc44ce09153c259 basepoints/gcc-7-3351-g56b653f */
+#define NB_COUNTERS 9
+#elif GCC_VERSION_10 <= __GCC_VERSION
+/* e37333bad7b7df7fd9d2e5165f61c2a68b57a30d basepoints/gcc-10-929-ge37333b */
+#define NB_COUNTERS 8
+#endif
+
+enum { CTR_ARCS = 0,
+       CTR_V_INTERVAL,
+       CTR_V_POW2,
+       CTR_V_TOPN,
+       CTR_V_INDIR,
+       CTR_AVERAGE,
+       CTR_IOR,
+       TIME_PROFILER,
+       START_CTR_SUPPORTED = CTR_ARCS,
+       LAST_CTR_SUPPORTED = CTR_ARCS + 1,
+};
+
+static const uint32_t ctr_tags[LAST_CTR_SUPPORTED] = {
+	[CTR_ARCS] = 0x01a10000,
+};
+
+/*
+ * Structure create by gcc to hold code coverade data of an object
+ *
+ * __gcov_ctr_info (func gcc/gcc/coverage.c:build_fn_info_type):
+ *   ctr_info::num: unsigned
+ *   ctr_info::values: <int size> *
+ *
+ * __gcov_fn_info (func build_fn_info_type):
+ *   key: const __gcov_info *
+ *   ident: unsigned
+ *   lineno_checksum: unsigned
+ *   cfg checksum: unsigned
+ *   counters: __gcov_ctr_info[number_of_gcov_counters_in_the_fn - 1]
+ */
+struct func_data {
+	/* Pointer to object containing the function */
+	const struct object_data *object;
+	/* Identifier */
+	uint32_t id;
+	/* Checksum of the line */
+	uint32_t line_chksum;
+	/* Checksum of the configuration */
+	uint32_t cfg_chksum;
+	/* Array of counters data for each type of counter */
+	struct ctr_data {
+		/* Number of counters of the type */
+		uint32_t nb_ctrs;
+		/* Array of counters value */
+		uint64_t *ctrs;
+	} ctrs_data[];
+};
+
+typedef void (*merge_fn)(uint64_t *ctrs, uint32_t nb_ctrs);
+
+/*
+ * Structure create by gcc to hold code coverade data of an object
+ *
+ * __gcov_info (func gcc/gcc/coverage.c:build_fn_info):
+ *   Version ident: unsigned
+ *   next pointer: const __gcov_info *
+ *   stamp: unsigned
+ *   Filename: const char *
+ *   merge fn array: void(*)(<int size>, unsigned)[number of gcov counters]
+ *   n_functions: unsigned
+ *   function_info pointer pointer: const
+ */
+struct object_data {
+	/* version of gcc */
+	uint32_t version;
+	/* pointer for a liked list */
+	struct object_data *next;
+	/* Date of creation */
+	uint32_t timestamp;
+	/* Name of the object */
+	const char *objectname;
+	/* Functions to merge the counters */
+	void (*merge_fn[NB_COUNTERS])(uint64_t *ctrs, uint32_t nb_ctrs);
+	/* Number of functions in the object */
+	uint32_t nb_functions;
+	/* Array of pointers on function data */
+	const struct func_data **arr_func_data;
+};
+
+/* Structure to hold data for the library */
+static struct internal_data {
+	/* Version of gcc */
+	uint32_t gcc_version;
+	/* Size of dump buffer, store the biggest size required */
+	uint32_t dump_buf_size;
+	/* Buffer to dump */
+	void *dump_buf;
+	/* List of all the objects covered */
+	struct object_data *linked_list_objects;
+	/* If a dry run is required */
+	bool dry_run_done;
+} s_int_data;
```

Comment:
It is to mark the structure is static helping when using it in a function (as it is not declared in the function) to find it.
It can be removed if it is too ugly and not useful enough.

Reply:
Please remove it then, feel free to spell it out as `internal_data` instead. We tend to use longer names for global (static or not) variables.